### PR TITLE
amp-analytics new triggers: touch, scroll

### PR DESCRIPTION
**Pull request by jonkeller**
Implements touch trigger, whitelists scroll trigger

### CODE REVIEW

File: extensions/amp-analytics/0.1/instrumentation.js [hunk 1]
```diff
@@ -124,6 +152,11 @@ const ALLOWED_IN_EMBED = [
   AnalyticsEventType.CLICK,
   AnalyticsEventType.TIMER,
   AnalyticsEventType.HIDDEN,
+  AnalyticsEventType.SCROLL,
```
Comment (jonkeller) [inline extensions/amp-analytics/0.1/instrumentation.js]: ^^ This is all I had to change for scroll to work for amp-analytics

File: extensions/amp-ad-network-fake-impl/0.1/data/fake_amp_ad_with_iframe_transport.html [hunk 2]
```diff
@@ -41,6 +69,38 @@
         }
       </script>
     </amp-analytics>
+    <amp-ad-exit id='exit-api'>
```
Comment (jonkeller) [inline extensions/amp-ad-network-fake-impl/0.1/data/fake_amp_ad_with_iframe_transport.html]: Inclusion of amp-ad-exit isn't really necessary here, but will save me a merge conflict with the other PR I'm currently working on.

File: extensions/amp-analytics/0.1/events.js [hunk 3]
```diff
@@ -407,6 +407,117 @@ export class IniLoadTracker extends EventTracker {
...
 }
 
+/**
+ * Tracks touch events
+ * @abstract
+ */
+class TouchEventTracker extends EventTracker {
+  /**
+   * @param {!./analytics-root.AnalyticsRoot} root
+   * @param {string} touchEventType 'touchstart', 'touchmove', 'touchend', or
```
Comment (taymonbeal) [inline extensions/amp-analytics/0.1/events.js]: Can this be an enum? Or is there an API compatibility issue?

File: extensions/amp-analytics/0.1/events.js [hunk 4]
```diff
@@ -407,6 +407,117 @@ export class IniLoadTracker extends EventTracker {
...
+   * 'touchcancel'
```
Comment (taymonbeal) [inline extensions/amp-analytics/0.1/events.js]: Indent four spaces

File: extensions/amp-analytics/0.1/events.js [hunk 5]
```diff
@@ -407,6 +407,117 @@ export class IniLoadTracker extends EventTracker {
...
+    dev().assert(touchEventType == 'touchstart' ||
+        touchEventType == 'touchmove' || touchEventType == 'touchend' ||
+        touchEventType == 'touchcancel',
+        'touchEventType unsupported: ' + touchEventType);
+    this.touchEventType_ = touchEventType;
+
+    /** @private {!Observable<!Event>} */
+    this.touchObservable_ = new Observable();
+
+    /** @private @const */
```
Comment (taymonbeal) [inline extensions/amp-analytics/0.1/events.js]: Type?

File: extensions/amp-analytics/0.1/events.js [hunk 6]
```diff
@@ -407,6 +407,117 @@ export class IniLoadTracker extends EventTracker {
...
+    super(root);
+
+    dev().assert(touchEventType == 'touchstart' ||
+        touchEventType == 'touchmove' || touchEventType == 'touchend' ||
+        touchEventType == 'touchcancel',
+        'touchEventType unsupported: ' + touchEventType);
+    this.touchEventType_ = touchEventType;
+
+    /** @private {!Observable<!Event>} */
+    this.touchObservable_ = new Observable();
```
Comment (taymonbeal) [inline extensions/amp-analytics/0.1/events.js]: Should this be `@const`?

File: extensions/amp-analytics/0.1/events.js [hunk 7]
```diff
@@ -407,6 +407,117 @@ export class IniLoadTracker extends EventTracker {
   }
 }
 
+/**
+ * Tracks touch events
+ * @abstract
+ */
+class TouchEventTracker extends EventTracker {
```
Comment (jonkeller) [inline extensions/amp-analytics/0.1/events.js]: Mostly copy-paste from ClickEventTracker earlier in this file

File: extensions/amp-analytics/0.1/events.js (hunk 6 above)
Comment (jonkeller) [inline extensions/amp-analytics/0.1/events.js]: Added code to dispose() to set it to null, which precludes const

File: extensions/amp-analytics/0.1/events.js (hunk 4 above)
Comment (jonkeller) [inline extensions/amp-analytics/0.1/events.js]: Shortened description, so moot now

File: extensions/amp-analytics/0.1/events.js (hunk 3 above)
Comment (jonkeller) [inline extensions/amp-analytics/0.1/events.js]: Done

File: extensions/amp-analytics/0.1/events.js (hunk 5 above)
Comment (jonkeller) [inline extensions/amp-analytics/0.1/events.js]: Done

File: extensions/amp-analytics/0.1/events.js [hunk 8]
```diff
@@ -407,29 +409,33 @@ export class IniLoadTracker extends EventTracker {
...
+   * @param {!TouchEventType} touchEventType Which type of touch event to track
    */
   constructor(root, touchEventType) {
     super(root);
 
-    dev().assert(touchEventType == 'touchstart' ||
-        touchEventType == 'touchmove' || touchEventType == 'touchend' ||
-        touchEventType == 'touchcancel',
-        'touchEventType unsupported: ' + touchEventType);
+    /** @const @private {TouchEventType} */
```
Comment (taymonbeal) [inline extensions/amp-analytics/0.1/events.js]: Need `!`

File: extensions/amp-analytics/0.1/events.js [hunk 9]
```diff
@@ -240,10 +240,10 @@ export class ClickEventTracker extends EventTracker {
   constructor(root) {
     super(root);
 
-    /** @private {!Observable<!Event>} */
+    /** @private {?Observable<!Event>} */
     this.clickObservable_ = new Observable();
 
-    /** @private @const */
+    /** @private {?Function} */
```
Comment (taymonbeal) [inline extensions/amp-analytics/0.1/events.js]: `?function(!Event)`

File: extensions/amp-analytics/0.1/events.js [hunk 10]
```diff
@@ -407,6 +409,123 @@ export class IniLoadTracker extends EventTracker {
...
+  constructor(root, touchEventType) {
+    super(root);
+
+    /** @const @private {TouchEventType} */
+    this.touchEventType_ = touchEventType;
+
+    /** @private {?Observable<!Event>} */
+    this.touchObservable_ = new Observable();
+
+    /** @private {?Function} */
```
Comment (taymonbeal) [inline extensions/amp-analytics/0.1/events.js]: `?function(!Event)`

File: extensions/amp-analytics/0.1/events.js [hunk 11]
```diff
@@ -407,6 +409,123 @@ export class IniLoadTracker extends EventTracker {
...
+ */
+class TouchEventTracker extends EventTracker {
+  /**
+   * @param {!./analytics-root.AnalyticsRoot} root
+   * @param {!TouchEventType} touchEventType Which type of touch event to track
+   */
+  constructor(root, touchEventType) {
+    super(root);
+
+    /** @const @private {TouchEventType} */
```
Comment (taymonbeal) [inline extensions/amp-analytics/0.1/events.js]: `!TouchEventType`

**Comment by jonkeller** [comment]
@lannka We are requesting a meeting with the vendor soon, and I will be sure to ask about granularity of events there.
The "touchstart" and "touchend" events will fire when a finger touches down onto the surface, or leaves it, so I don't think those two will be very spammy. On the other hand, "touchmove" fires any time a finger(s) moves - in my testing I can get this to happen ~100x/second. Do you have more information on the external effort to batch/queue events that you mentioned previously? Sounds like this would be a good candidate, if that is ready. If not, I can add my old throttling code back in, if you like.

**Comment by lannka** [comment]
@jonkeller yep, let's make sure sending 100x/second touch move events is useful for them. On the other side, this makes this trigger useless in other transports scenario. It would be nice if pub/vendor can also use this new trigger to collect user touch events, which was one of the reasons why we decided to make the spam signal trigger part of amp-analytics.

If possible, I would propose to let amp-analytics observe a whole touch event, and trigger only once at the end, and sending a event proto like:
```
{
  touchstartPos: {x, y},
  touchendPos: {x, y},
  touchDurationInMs: 120,
}
```
Details about the proto can be discussed with vendor. In this way, no throttling or batching would be needed for this type of trigger.

**Throttling & batching**
Throttling or batching would still be needed for other trigger like "scroll".
My proposal was to throttle at event trigger level. (See the original discussion #10367). Noticing that different browsers already throttle "scroll" events in different ways, AMP runtime can just apply a further protection on it (e.g. 1/100ms).

**Comment by jonkeller** [comment]
@lannka SGTM. I definitely do want to include touch coords in the event string (or allow vendor to request them via variable substitution, etc.). I'll update once we've talked to the vendor more. Thanks.

**Comment by jonkeller** [comment]
FYI, we met with the vendor on Thursday of last week. They are preparing a list of requirements, which they shall send to us this week.

**Comment by jonkeller** [comment]
Discussed this in mtg today. Will pick back up in 1Q2018.

**Pull request closed without merging**